akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,48 +4,139 @@
+    val ioResultPromise = Promise[IOResult]()
 
-  override protected def label: String = s"FileSource($f, $chunkSize)"
+    val logic = new GraphStageLogic(shape) with OutHandler {
+      handler ⇒
+      val buffer = ByteBuffer.allocate(chunkSize)
+      val maxReadAhead = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+      var channel: FileChannel = _
+      var position = startPosition
+      var chunkCallback: Try[Int] ⇒ Unit = _
+      var eofEncountered = false
+      var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+      setHandler(out, this)
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+
+          channel = FileChannel.open(path, StandardOpenOption.READ)
+          channel.position(position)
+        } catch {
+          case ex: Exception ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+            throw ex
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (availableChunks.size < maxReadAhead && !eofEncountered)
+          availableChunks = readAhead(maxReadAhead, availableChunks)
+        //if already read something and try
+        if (availableChunks.nonEmpty) {
+          emitMultiple(out, availableChunks.iterator,
+            () ⇒ if (eofEncountered) success() else setHandler(out, handler)
+          )
+          availableChunks = Vector.empty[ByteString]
+        } else if (eofEncountered) success()
+      }
+
+      private def success(): Unit = {
+        completeStage()
+        ioResultPromise.trySuccess(IOResult(position, Success(Done)))
+      }
+
+      /** BLOCKING I/O READ */
+      @tailrec def readAhead(maxChunks: Int, chunks: Vector[ByteString]): Vector[ByteString] =
+        if (chunks.size < maxChunks && !eofEncountered) {
+          val readBytes = try channel.read(buffer, position) catch {
+            case NonFatal(ex) ⇒
+              failStage(ex)
+              ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+              throw ex
+          }
+
+          if (readBytes > 0) {
+            buffer.flip()
+            position += readBytes
+            val newChunks = chunks :+ ByteString.fromByteBuffer(buffer)
+            buffer.clear()
+
+            if (readBytes < chunkSize) {
+              eofEncountered = true
+              newChunks
+            } else readAhead(maxChunks, newChunks)
+          } else {
+            eofEncountered = true
+            chunks
+          }
+        } else chunks
+
+      override def onDownstreamFinish(): Unit = success()
+
+      override def postStop(): Unit = {
+        if ((channel ne null) && channel.isOpen) channel.close()
```

Comment:
I think I had a comment about trying to complete the `ioResultPromise` from `postStop` also. Perhaps I missed your reply. Shouldn't it? It can happen when abrupt termination, e.g. materializer shutdown.

Reply:
Ah, I got your point. Will add this. I think test will  use mapMaterializelue to emulate materialization problems